fix: correct upstream field for migrated libraries

ormsbee opened the pull request:
This is to fix an issue in the following common migration situation:

1. An existing course references content in a legacy content library.
2. The legacy content library is migrated to the new library system.
3. The user clicks on "Update reference" from the Randomized Content Block in the course.

This action is supposed to update the children of the LibraryContentBlock (usually ProblemBlocks) so that the "upstream" attribute is set to point at the UsageKeys of the content in the new libraries they were migrated to. What was happening instead was that the upstream entries for these child blocks were left blank, breaking the upstream/sync connection and making it so that the courses did not receive any updates from the migrated libraries.

There were two issues:

1. get_forwarding_for_blocks() was being called with the child UsageKeys in the course, when it should have been called with the v1 library usage keys instead (since those are the things being forwarded).
2. We were checking that the target_key was a v2 Library key, but really the upstream target_key is supposed to be a LibraryUsageLocatorV2, i.e. the key of the specific piece of content, not the library it ended up in.

----

Fixes #37802

I still need to write a test for this, and test a bit more to make sure it's otherwise functioning as expected.

ormsbee committed:
fix: correct upstream field for migrated libraries

This is to fix an issue in the following common migration situation:

1. An existing course references content in a legacy content library.
2. The legacy content library is migrated to the new library system.
3. The user clicks on "Update reference" from the Randomized Content
   Block in the course.

This action is supposed to update the children of the
LibraryContentBlock (usually ProblemBlocks) so that the "upstream"
attribute is set to point at the UsageKeys of the content in the new
libraries they were migrated to. What was happening instead was that the
upstream entries for these child blocks were left blank, breaking the
upstream/sync connection and making it so that the courses did not
receive any updates from the migrated libraries.

There were two issues:

1. get_forwarding_for_blocks() was being called with the child UsageKeys
   in the course, when it should have been called with the v1 library
   usage keys instead (since those are the things being forwarded).
2. We were checking that the target_key was a v2 Library key, but really
   the upstream target_key is supposed to be a LibraryUsageLocatorV2,
   i.e. the key of the specific piece of content, not the library it
   ended up in.

kdmccormick reviewed (commented):
The reasoning and the code change both look solid to me. I haven't manually tested yet but I'll do that shortly. I'll also make a unit test for this.

kdmccormick committed:
test: Minimal unit test for upstream field migration bugfix

Although there were unit tests for the migration of
legacy content libraries, there were not any unit tests
for the migration of legacy library *blocks*.

This adds a minimal test, which would have caught the bug which
we fix in the previous commit.
It would be good to add more comprephensive testing unit testing
for this part of the migration flow.

kdmccormick reviewed (approved):
Approving @ormsbee 's commit 👍🏻 

@ormsbee , could you review my commit, and approve and merge if it looks good? If there are just minor tweaks you have feel free to go ahead fix them.

I have not manually tested this (my tutor image build kept failing today) but it sounds like you have @ormsbee , and we will be testing on the sandbox as well.

kdmccormick commented on xmodule/tests/test_library_content.py:
```diff
@@ -770,6 +768,42 @@ def setUp(self):
+        This tests only the simplest state: The source lib has been migrated with forwarding, exactly once,
+        and the LC block has also been migrated.
+
+        It would be good to also test more cases, including:
```
```suggestion
        TODO(https://github.com/openedx/edx-platform/issues/37837):
        It would be good to also test more cases, including:
```

ormsbee committed:
temp: update xmodule/tests/test_library_content.py

Co-authored-by: Kyle McCormick <kyle@kylemccormick.me>
